File: core/deterministic/src/motion_estimation/motion_estimation.rs
use kdam::tqdm;
use opencv::prelude::Mat;

use crate::{
    video::Video, config_parameters::{ConfigParameters},
};

use super::{block_matching::BlockMatching};

pub struct MotionEstimation<'a> {
    pub block_matching: BlockMatching<'a>,
    video: &'a Video<'a>,
    config_parameters: &'a ConfigParameters,
}

impl<'a> MotionEstimation<'a> {
    pub fn new(video: &'a mut Video, config_parameters: &'a ConfigParameters) -> Self {
        Self {
            block_matching: BlockMatching::new(
                config_parameters.dfd.clone(),
                config_parameters.block_size.clone(),
                config_parameters.search_range.clone(),
                config_parameters.motion_intensity.clone(),
                config_parameters,
            ),
            video: video,
            config_parameters: config_parameters,
        }
    }

    pub fn demo(&mut self, anchor_index: usize, target_index: usize) -> (Option<(f32, f32)>, Option<Mat>, Option<Mat>, Option<Mat>) {
        if self.config_parameters.frames_print_debug {
            let anchor_frame = self.video.frames_inp[anchor_index].clone();
            let target_frame = self.video.frames_inp[target_index].clone();

            return self.block_matching.step(anchor_frame.clone(), target_frame.clone());
        }
        (None, None, None, None)
    }

    pub fn video_processing(&mut self) -> (Vec<(f32, f32)>, Vec<Option<Mat>>, Vec<Option<Mat>>, Vec<Option<Mat>>){
        // Video Motion Estimation Processing
        println!("Motion Estimation (Block Matching - Three Step Search) processing...");
        let mut global_motion_vectors: Vec<(f32, f32)> = vec![];
        let mut frame_anchor_p_vec: Vec<Option<Mat>> = vec![];
        let mut frame_motion_field_vec: Vec<Option<Mat>> = vec![];
        let mut frame_global_motion_vec: Vec<Option<Mat>> = vec![];
        for f in tqdm!(0..self.video.total_frame.unwrap() - 1) {
            let anchor =  self.video.frames_inp[f as usize].clone();
            let target = self.video.frames_inp[(f + 1) as usize].clone();

            if self.config_parameters.frames_print_debug {
                let (global_motion_vec, frame_anchor_p, frame_motion_field, frame_global_motion_vector) = self.block_matching.step(anchor.clone(), target.clone());

                global_motion_vectors.push(global_motion_vec.unwrap());
                frame_anchor_p_vec.push(frame_anchor_p);
                frame_motion_field_vec.push(frame_motion_field);
                frame_global_motion_vec.push(frame_global_motion_vector);
            } else {
                let (global_motion_vec, _, _, _) = self.block_matching.step(anchor.clone(), target.clone());
                global_motion_vectors.push(global_motion_vec.unwrap());

            }
        }
        (global_motion_vectors, frame_anchor_p_vec, frame_motion_field_vec, frame_global_motion_vec)
    }
}

File: core/deterministic/python-core/src/motion_estimation/motion_estimation.py
import cv2
from fastapi import WebSocket, WebSocketDisconnect
from tqdm import tqdm
from .block_matching import BlockMatching
from config.config_video import ConfigVideoParameters
from src.request_handler.json_encoder import JsonEncoder
class MotionEstimation:
    def __init__(self, config_parameters: ConfigVideoParameters):
        self.config_parameters = config_parameters
        
    async def video_processing(self, frames, websocket: WebSocket, update_step_code = 'me', compare_message = ""):
        message = "Motion Estimation (Block Matching - Three Step Search) processing " + compare_message + "..."
        print(message)
        await websocket.send_json(JsonEncoder.init_motion_estimation_json(message, state=update_step_code))

        global_motion_vectors = []
        frame_anchor_p_vec = []
        frame_motion_field_vec = []
        frame_global_motion_vec = []
        block_matching = BlockMatching(self.config_parameters)
        
        _range = range(len(frames) - 1)
        total = _range[-1]
        for f in tqdm(_range):
            anchor =  cv2.cvtColor(frames[f], cv2.COLOR_BGR2GRAY)
            target = cv2.cvtColor(frames[f + 1], cv2.COLOR_BGR2GRAY)

            if self.config_parameters.demo:
                global_motion_vec, frame_anchor_p, frame_motion_field, frame_global_motion_vector = block_matching.step(anchor, target)
                
                global_motion_vectors.append(global_motion_vec)
                frame_anchor_p_vec.append(frame_anchor_p)
                frame_motion_field_vec.append(frame_motion_field)
                frame_global_motion_vec.append(frame_global_motion_vector)
                
            else:
                global_motion_vec, _, _, _ = block_matching.step(anchor, target)
                global_motion_vectors.append(global_motion_vec)
        
            await websocket.send_json(JsonEncoder.update_step_json(update_step_code, f, total))
            try:
                await websocket.receive_text()
            except WebSocketDisconnect:              
                raise

        return global_motion_vectors, frame_anchor_p_vec, frame_motion_field_vec, frame_global_motion_vec
